### PR TITLE
Implement css property contain-intrinsic-size

### PR DESCRIPTION
**Pull request by cathiechen**
#### 7ed83ab2929f7a2be55fb40888fea7ff02909f56
<pre>
Implement css property contain-intrinsic-size
<a href="https://bugs.webkit.org/show_bug.cgi?id=238867">https://bugs.webkit.org/show_bug.cgi?id=238867</a>

Reviewed by Alan Bujtas.

This patch adds support for contain-intrinsic-size value &quot;none | &lt;length&gt;&quot;. The value &quot;auto &amp;&amp; &lt;length&gt;&quot;
is not supported, because per [1] it depends on &quot;content-visibility:auto&quot; which is not supported yet.
Per [2], the properties allow elements with size containment to specify an explicit intrinsic inner size,
which is determined in RenderBox. These values will affect the intrinsic width calculation in computeIntrinsicLogicalWidths.
As to intrinsic height, the value will be overridden at the point after layout children and before logical height calculation.
For grid layout, we need to use explicit intrinsic inner size to calculate auto repetition, if width/height is auto.

[1] <a href="https://github.com/w3c/csswg-drafts/issues/6308">https://github.com/w3c/csswg-drafts/issues/6308</a>
[2] <a href="https://www.w3.org/TR/css-sizing-4/#intrinsic-size-override">https://www.w3.org/TR/css-sizing-4/#intrinsic-size-override</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-005-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-009-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-029-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-030-expected.txt: The three failed cases are related to css scrollbar-gutter.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-031-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-032-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::GridTrackSizingAlgorithm::computeGridContainerIntrinsicSizes): Set explicitIntrinsicInnerLogicalSize to m_minContentSize and m_maxContentSize.
(WebCore::GridTrackSizingAlgorithm::run): Don&apos;t skip stretching tracks for size containment if there is explicitIntrinsicInnerLogicalSize.
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderBox.cpp:
(WebCore::RenderBox::updateLogicalHeight): Set explicitIntrinsicHeight to logicalHeight if it exists.
(WebCore::RenderBox::computeLogicalHeightWithoutLayout const): We should take account of explicitIntrinsicInnerLogicalHeight here.
(WebCore::RenderBox::explicitIntrinsicInnerWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerHeight const):
* Source/WebCore/rendering/RenderBox.h:
(WebCore::RenderBox::explicitIntrinsicInnerLogicalWidth const):
(WebCore::RenderBox::explicitIntrinsicInnerLogicalHeight const):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderFileUploadControl.cpp:
(WebCore::RenderFileUploadControl::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutBlock): Should use explicitIntrinsicInnerHeigth as the definite size if exists.
(WebCore::RenderGrid::computeIntrinsicLogicalWidths const): Don&apos;t add gutterSize if it&apos;s explicitIntrinsicInnerLogicalSize.
(WebCore::RenderGrid::computeTrackSizesForIndefiniteSize const): Ditto.
(WebCore::RenderGrid::explicitIntrinsicInnerLogicalSize const): Ditto.
(WebCore::RenderGrid::computeAutoRepeatTracksCount const): Use explicitIntrinsicInnerLogicalSize as availableSize if the size of grid container is not specified.
(WebCore::RenderGrid::computeEmptyTracksForAutoRepeat const): The tracks of grid container with size containment are not empty if hasExplicitIntrinsicInnerLogicalSize.
* Source/WebCore/rendering/RenderGrid.h:
* Source/WebCore/rendering/RenderListBox.cpp:
(WebCore::RenderListBox::computeIntrinsicLogicalWidths const): Ditto.
(WebCore::RenderListBox::computeLogicalHeight const): Use explicitIntrinsicInnerLogicalHeight as estimated height not the sum of item height.
* Source/WebCore/rendering/RenderMenuList.cpp:
(RenderMenuList::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderReplaced.cpp:
(WebCore::RenderReplaced::computeIntrinsicRatioInformation const): Don&apos;t let size containment effect aspect ratio.
* Source/WebCore/rendering/RenderReplaced.h: Return explicitIntrinsicInnerSize if exits.
* Source/WebCore/rendering/RenderSlider.cpp:
(WebCore::RenderSlider::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderTextControl.cpp:
(WebCore::RenderTextControl::computeIntrinsicLogicalWidths const): Ditto.
* Source/WebCore/rendering/RenderVideo.cpp:
(WebCore::RenderVideo::calculateIntrinsicSize):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::rareNonInheritedDataChangeRequiresLayout): The changes of contain-intrinsic-size should require layout.

Canonical link: <a href="https://commits.webkit.org/255971@main">https://commits.webkit.org/255971@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5b0828900abb00fe80b37899e52d427b18a495b7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94190 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3382 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/103813 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164144 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98179 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3401 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31596 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86502 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/99834 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/99845 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/2440 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/80608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29461 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84364 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72410 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/37978 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/35863 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4129 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/39738 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/80608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41680 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/24717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

**Comment by cathiechen** [comment]
Hi @smfr @nt1m !
This is the implementation of contain-intrinsic-size, rebased the latest code.

File: Source/WebCore/rendering/RenderBox.h [hunk 1]
```diff
@@ -790,6 +804,9 @@ override;
     
     void applyTopLeftLocationOffsetWithFlipping(LayoutPoint&) const;
 
+    bool hasExplicitIntrinsicInnerWidth() const;
+    bool hasExplicitIntrinsicInnerHeight() const;
```
Comment (rwlbuis) [inline Source/WebCore/rendering/RenderBox.h]: It looks like these two methods are not needed, assuming they will only be used internally in RenderBox.cpp.

File: Source/WebCore/rendering/RenderFlexibleBox.cpp [hunk 2]
```diff
@@ -98,8 +98,12 @@ void RenderFlexibleBox::computeIntrinsicLogicalWidths(LayoutUnit& minLogicalWidt
         minLogicalWidth += scrollbarWidth;
     };
 
-    auto shouldIgnoreFlexItemContentForLogicalWidth = shouldApplySizeOrInlineSizeContainment();
-    if (shouldIgnoreFlexItemContentForLogicalWidth) {
+    bool isSizeContainment = shouldApplySizeContainment();
```
Comment (rwlbuis) [inline Source/WebCore/rendering/RenderFlexibleBox.cpp]: Better here is needsSizeContainment.

File: LayoutTests/TestExpectations [hunk 3]
```diff
@@ -4729,34 +4729,7 @@ webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/
 webkit.org/b/214463 imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/floats-aspect-ratio-001.html [ ImageOnlyFailure ]
 
 # contain-intrinsic-size
```
Comment (rwlbuis) [inline LayoutTests/TestExpectations:4786]: Probably this comment can be removed too.

Comment (cathiechen) [reply LayoutTests/TestExpectations:4786]: Done

File: Source/WebCore/rendering/RenderFlexibleBox.cpp (hunk 2 above)
Comment (cathiechen) [reply Source/WebCore/rendering/RenderFlexibleBox.cpp]: Done

File: Source/WebCore/rendering/RenderBox.h (hunk 1 above)
Comment (cathiechen) [reply Source/WebCore/rendering/RenderBox.h]: Done, thanks!

File: Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp [hunk 4]
```diff
@@ -202,6 +202,8 @@ void RenderDeprecatedFlexibleBox::computeIntrinsicLogicalWidths(LayoutUnit& minL
     };
 
     if (shouldApplySizeContainment()) {
+        if (auto width = explicitIntrinsicInnerLogicalWidth())
+            minLogicalWidth = maxLogicalWidth = width.value();
```
Comment (rwlbuis) [inline Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp]: These two lines happen a lot, could we have a shared method?

File: Source/WebCore/rendering/RenderVideo.cpp [hunk 5]
```diff
@@ -117,8 +117,14 @@ bool RenderVideo::updateIntrinsicSize()
...
 {
-    if (shouldApplySizeContainment())
-        return LayoutSize();
+    if (shouldApplySizeContainment()) {
+        LayoutSize size;
+        if (auto width = explicitIntrinsicInnerWidth())
+            size.setWidth(width.value());
+        if (auto height = explicitIntrinsicInnerHeight())
+            size.setHeight(height.value());
+        return size;
```
Comment (rwlbuis) [inline Source/WebCore/rendering/RenderVideo.cpp]: Is this needed since we have RenderReplaced:: intrinsicSize? If yes, can we share the code?

Comment (cathiechen) [reply Source/WebCore/rendering/RenderVideo.cpp]: I'm not quite sure what does this mean?

Comment (rwlbuis) [reply Source/WebCore/rendering/RenderVideo.cpp]: RenderVideo is a RenderReplaced, so why is the RenderReplaced:: intrinsicSize change not enough?

File: Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp (hunk 4 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp]: Ok, I misread, I thought we were setting member vars here. I don't think we can improve much here.

File: Source/WebCore/rendering/RenderVideo.cpp (hunk 5 above)
Comment (cathiechen) [reply Source/WebCore/rendering/RenderVideo.cpp]: OK, makes sense. I don't quite remember, is this to solve some issue or not.  Let me try to remove it.

Comment (rwlbuis) [reply Source/WebCore/rendering/RenderVideo.cpp]: It seems to work! Can you adjust the commit message to explain this better?:
 (WebCore::RenderVideo::calculateIntrinsicSize): Ditto.
I think github allows to edit the commit message by hand...

Comment (cathiechen) [reply Source/WebCore/rendering/RenderVideo.cpp]: Sure

**Comment by rwlbuis** [comment]
The patch now looks good to me!

**Comment by Loirooriol** [comment]
Consider importing WPT, new tests have been added during the implementation in Gecko.

**Comment by rwlbuis** [comment]
> Consider importing WPT, new tests have been added during the implementation in Gecko.

Good idea, working on that here https://bugs.webkit.org/show_bug.cgi?id=245271.

File: LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-030-expected.txt [hunk 6]
```diff
@@ -1,62 +1,32 @@
 
 PASS .test 1
-FAIL .test 2 assert_equals:
-<div class="test small-contents overflow-auto cis-height" data-expected-client-width="0" data-expected-client-height="50" data-expected-scroll-width="40" data-expected-scroll-height="50"></div>
-clientHeight expected 50 but got 0
+PASS .test 2
 FAIL .test 3 assert_equals:
```
Comment (cathiechen) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-030-expected.txt:4]: The cases 3, 15 and 16 use `scrollbar-gutter: stable;` to turn on the scrollbar. `scrollbar-gutter` is not supported on WebKit yet, so these cases are failing now.
Left a comment on bug: 167335

File: LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-032-expected.txt [hunk 7]
```diff
@@ -20,11 +20,11 @@ FAIL .test 6 assert_equals:
 clientWidth expected 0 but got 34
 FAIL .test 7 assert_equals:
 <select class="test ciw-100 cih-none" data-expected-client-width="100" data-expected-client-height="18"></select>
```
Comment (cathiechen) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-032-expected.txt]: The failures in this test are related to https://bugs.webkit.org/show_bug.cgi?id=246338

File: LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt [hunk 8]
```diff
@@ -1,128 +1,65 @@
...
-<svg class="test cis-inline vertical" data-expected-client-height="100" data-expected-client-width="0"></svg>
-clientHeight expected 100 but got 0
-FAIL .test 24 assert_equals:
-<svg class="test cis-both vertical" data-expected-client-height="100" data-expected-client-width="50"></svg>
-clientWidth expected 50 but got 0
+PASS .test 22
+PASS .test 23
+PASS .test 24
 FAIL .test 25 assert_equals:
 <canvas class="test cis-none" width="40" height="20" data-expected-client-width="0" data-expected-client-height="0"></canvas>
```
Comment (cathiechen) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/contain-intrinsic-size-logical-003-expected.txt:27]: These are related to https://bugs.webkit.org/show_bug.cgi?id=246283

File: LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt [hunk 9]
```diff
@@ -1,3 +1,3 @@
 
-FAIL contain-intrinsic-size: auto assert_equals: expected 1 but got 0
+FAIL contain-intrinsic-size: auto assert_equals: expected 50 but got 1
```
Comment (cathiechen) [inline LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/contain-intrinsic-size/auto-003-expected.txt:2]: `contain-intrinsic-size: auto && <length>` is not supported yet, it depends on `content-visibility`. https://bugs.webkit.org/show_bug.cgi?id=239450

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp [hunk 10]
```diff
@@ -780,6 +780,13 @@ double GridTrackSizingAlgorithm::findFrUnitSize(const GridSpan& tracksSpan, Layo
 
 void GridTrackSizingAlgorithm::computeGridContainerIntrinsicSizes()
 {
+    if (m_direction == ForColumns && m_strategy->isComputingSizeContainment()) {
+        if (auto size = m_renderGrid->explicitIntrinsicInnerLogicalSize(m_direction)) {
+            m_minContentSize = m_maxContentSize = size.value();
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: use the WebKit style here, please (and at all the other places)
m_minContentSize = size.value();
m_maxContentSize = size.value();

File: Source/WebCore/rendering/RenderBox.cpp [hunk 11]
```diff
@@ -5555,4 +5563,26 @@ bool RenderBox::shouldIgnoreMinMaxSizes() const
...
 
+std::optional<LayoutUnit> RenderBox::explicitIntrinsicInnerWidth() const
+{
+    ASSERT(shouldApplySizeContainment());
+    if (style().containIntrinsicWidthType() == ContainIntrinsicSizeType::None)
+        return std::nullopt;
+
+    auto width = style().containIntrinsicWidth();
+    ASSERT(width.has_value());
+    return std::optional<LayoutUnit>(width.value().value());
```
Comment (alanbaradlay) [inline Source/WebCore/rendering/RenderBox.cpp]: you could just do (same below) 
return { width->value() }

**Review by alanbaradlay — APPROVED** [review]
While I can't tell if this covers all the cases (and whether it's missing some pieces), we should land it and watch for the fallout.

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp (hunk 10 above)
Comment (cathiechen) [reply Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp]: Done, thanks!

File: Source/WebCore/rendering/RenderBox.cpp (hunk 11 above)
Comment (cathiechen) [reply Source/WebCore/rendering/RenderBox.cpp]: Ah, yes, it looks better!
But compiler complains: `error: chosen constructor is explicit in copy-initialization`.
So I use `return std::optional<LayoutUnit>{ width->value() };` in the new patch

**Comment by cathiechen** [comment]
> While I can't tell if this covers all the cases (and whether it's missing some pieces), we should land it and watch for the fallout.

Thanks for the review! It made my day\o/

**Pull request merged**